### PR TITLE
Graph: refactored loader to use TypeScript

### PR DESCRIPTION
**Pull request by stephanwlee**
graph-dashboard is complex enough to benefit from good typing and
currently it is not being utilized well some in Closure and others in
TypeScript.

By refactoring, we were able to uncover bad typing on loader when using
tf.graph.hierarchy.

Note that other refactors will follow (e.g., more typing)

Tested by running TensorBoard

### CODE REVIEW

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+      msg: ''
+    });
+    var tracker = tf.graph.util.getTracker(this);
+    var hierarchyParams = {
+      verifyTemplate: true,
+      // If a set of numbered op nodes has at least this number of nodes
+      // then group them into a series node.
+      seriesNodeMinSize: 5,
+      // A map of series node names to series grouping settings, to indicate
+      // if a series is to be rendered as grouped or ungrouqped.
```
Comment (wchargin) [inline tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: new typo: *ungrou**q**ped*

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+      // Starts out empty which allows the renderer to decide which series
+      // are initially rendered grouped and which aren't.
+      seriesMap: {},
+      rankDirection: 'BT',
```
Comment (wchargin) [inline tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Lines 180–181 are new in this commit. Why?

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    if (!file) {
+      return;
+    }
+
+    // Clear out the value of the file chooser. This ensures that if the user
+    // selects the same file, we'll re-read it.
+    e.target.value = '';
+
+    // Clear stats about the previous graph.
+    this._setOutStats(null);
```
Comment (wchargin) [inline tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: This line is new in this commit. Why?

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        const shouldFetchGraph = !this._graphRunTag ||
+            this._graphRunTag.run !== run ||
+            this._graphRunTag.tag !== requiredOpGraphTag;
+        const maybeFetchGraphPromise = shouldFetchGraph ?
+            this._load({
+              run,
+              tag: requiredOpGraphTag,
+              type: tf.graph.SelectionType.OP_GRAPH,
+            }) : Promise.resolve();
+        const params = new URLSearchParams();
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Yes but TypeScript does not seem to know about it and complains.

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts (hunk 2 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Yeah and shrug. This component was consuming the API incorrectly and it was complaining about wrong type.

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts (hunk 3 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Must have mixed two commits by mistake. Will revert

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wchargin) [reply tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Hmm, interesting. [It works fine in the TypeScript playground][1], but
indeed it seems to fail in our configuration. No action required, then;
thanks.

[1]: http://www.typescriptlang.org/play/index.html#src=const%20tag%20%3D%20%22one%22%3B%0D%0Aconst%20run%20%3D%20%22two%22%3B%0D%0Anew%20URLSearchParams(%7B%20tag%2C%20run%20%7D)%3B

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts (hunk 2 above)
Comment (wchargin) [reply tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Got it; thanks!

File: tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    // Clear out the value of the file chooser. This ensures that if the user
+    // selects the same file, we'll re-read it.
+    target.value = '';
+
+    this._fetchAndConstructHierarchicalGraph(
+        null, file, overridingHierarchyParams);
+  },
+});
+
+}  // namespace tf.graph.loader
```
Comment (wchargin) [inline tensorboard/plugins/graph/tf_graph_loader/tf-graph-loader.ts]: Sorry, forgot to mention in first pass: missing newline at eof.

**Pull request merged**